admin_ui/src/modules/translations.ts

```diff
@@ -0,0 +1,107 @@
+import axios from "axios"
+
+import {
+    TranslationsListAPIResponse,
+    TranslationListItemAPI,
+    TranslationAPIResponse
+} from "@/interfaces"
+import i18n from "@/i18n"
+
+const DEFAULT_LANGUAGE_KEY = "piccoloAdminDefaultLanguage"
+
+/**
+ * Stores the user's default language in localStorage. If the user selects a
+ * language, then this becomes their new default language, otherwise it's
+ * defined by the API.
+ */
+const localStorageUtils = {
+    getDefaultLanguage: (): string | null => {
+        return localStorage.getItem(DEFAULT_LANGUAGE_KEY)
+    },
+    setDefaultLanguage: (value: string) => {
+        return localStorage.setItem(DEFAULT_LANGUAGE_KEY, value)
+    }
+}
+
+export default {
+    state: {
+        translations: null as string | TranslationListItemAPI[]
+    },
+    mutations: {
+        updateTranslations(state, value: object) {
+            state.translations = value
+        }
+    },
+    actions: {
+        async setupTranslations(context) {
+            // This fetches a list of all available translations (not the
+            // translations themselves):
+            const response = await axios.get<TranslationsListAPIResponse>(
+                `./public/translations/`
+            )
+            const data = response.data
+            context.commit("updateTranslations", data.translations)
+
+            const availableLanguageCodes = data.translations.map(
+                (translation) => {
+                    return translation.language_code.toLowerCase()
+                }
+            )
+
+            let defaultLanguage: string | null = null
+
+            // If the user has previously used the app, and set their default
+            // language, then load it from localStorage:
+            const storedLanguagePreference =
+                localStorageUtils.getDefaultLanguage()
+
+            if (
+                storedLanguagePreference &&
+                availableLanguageCodes.indexOf(storedLanguagePreference) != -1
+            ) {
+                defaultLanguage = storedLanguagePreference
```

Comment:
This should be `defaultLanguage = response.data.default_language_code` because otherwise we get the last selected language from the dropdown as the default, which is not the case. We should get the default language as we specified in `create_admin()`. That's why we need `default_language_code: str = "en"` because if it is set as `auto` during logout, we get a `404` error because `http://piccolo_admin/translations/auto` does not exist.

Reply:
Yeah, I intended it to the be like that. So the logic is:

 * If the default language is set to 'auto', get the user's language preference from the browser. If we don't have translations available for their language, it falls back to English.
 * If the default language is something other than 'auto', use that language by default.
 * When the user picks a different language from the dropdown, that becomes their default.

If the app is makin an API call to `/translations/auto` then that's definitely a bug though.

Reply:
I'm sorry I didn't realize that. Then forget my comments about it. Everything is fine and there are no bug because `/translations/auto` is called only if we change the `default_language_code` to `"en"`. If it is `"auto"` everything is fine.

Reply:
No worries - the logic changed quite a bit. It's more complex, but I think it covers all use cases.